storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -0,0 +1,420 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.storm.Config;
+import org.apache.storm.blobstore.BlobStore;
+import org.apache.storm.blobstore.ClientBlobStore;
+import org.apache.storm.daemon.Shutdownable;
+import org.apache.storm.daemon.supervisor.AdvancedFSOps;
+import org.apache.storm.daemon.supervisor.SupervisorUtils;
+import org.apache.storm.generated.StormTopology;
+import org.apache.storm.utils.ConfigUtils;
+import org.apache.storm.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * This is a wrapper around the Localizer class that provides the desired
+ * async interface to Slot.
+ * TODO once we have replaced the original supervisor merge this with
+ * Localizer and optimize them
+ */
+public class AsyncLocalizer implements ILocalizer, Shutdownable {
+    /**
+     * A future that has already completed.
+     */
+    private static class AllDoneFuture implements Future<Void> {
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return false;
+        }
+
+        @Override
+        public boolean isDone() {
+            return true;
+        }
+
+        @Override
+        public Void get() {
+            return null;
+        }
+
+        @Override
+        public Void get(long timeout, TimeUnit unit) {
+            return null;
+        }
+
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(AsyncLocalizer.class);
+
+    private final Localizer _localizer;
+    private final ExecutorService _execService;
+    private final boolean _isLocalMode;
+    private final Map<String, Object> _conf;
+    private final Map<String, LocalDownloadedResource> _basicPending;
+    private final Map<String, LocalDownloadedResource> _blobPending;
+    private final AdvancedFSOps _fsOps;
+
+    private class DownloadBaseBlobsDistributed implements Callable<Void> {
+        private final String _topologyId;
+        
+        public DownloadBaseBlobsDistributed(String topologyId) {
+            this._topologyId = topologyId;
+        }
+        
+        @Override
+        public Void call() throws Exception {
+            String stormroot = ConfigUtils.supervisorStormDistRoot(_conf, _topologyId);
+            File sr = new File(stormroot);
+            if (sr.exists()) {
+                if (!_fsOps.supportsAtomicDirectoryMove()) {
+                    LOG.warn("{} may have partially downloaded blobs, recovering", _topologyId);
+                    Utils.forceDelete(stormroot);
+                } else {
+                    LOG.warn("{} already downloaded blobs, skipping", _topologyId);
+                    return null;
+                }
+            }
```

Comment:
Let's make a helper method for this `if`, since we do the same thing for the DownloadBaseBlobsLocal.